program.py

```diff
@@ -7,1189 +7,369 @@
 
     tokenEndpoint = json.loads(discoveryUrl.content)["token_endpoint"]
+    tokenUrl = urlparse(tokenEndpoint)
+    # Validate URL
+    assert tokenUrl.scheme == 'https'
+    assert tokenUrl.geturl().startswith(endpoint["resource"])
 
     tokenInformation = requests.post(
-        tokenEndpoint,
-        data={"client_id": clientId,
-              "client_secret": clientSecret,
-              "grant_type": "client_credentials"},
-        verify=VERIFY_SSL)
+        tokenUrl.geturl(),
+        data={'client_id': endpoint["client-id"],
+              'client_secret': endpoint["client-secret"],
+              'grant_type': 'client_credentials'},
+        verify=endpoint["verify-ssl"])
 
     token = json.loads(tokenInformation.content)
 
     if token is None:
-        raise Exception("Failed to retrieve Token")
+        raise Exception('Failed to retrieve Token')
+
+    __expiration = float(token["expires_in"]) + time.time()
+    __token = token["access_token"]
+
+    # cache the results
+    endpoint["expiration"] = __expiration
+    endpoint["token"] = __token
 
-    __expiration = float(token['expires_in']) + time.time()
-    __token = token['access_token']
     return __token
 
 
 # ************************************************************************
-# Helper function: REQUIRED: wrapper function for sending an HTTPS message
+# REQUIRED: wrapper function for sending an HTTP message
 # ************************************************************************
 
-# Define a helper function to allow easily sending web request messages;
-# this function can later be customized to allow you to port this script to other languages.
-# All it does is take in a data object and a message type, and it sends an HTTPS
-# request to the target OMF endpoint
 
-def send_omf_message_to_endpoint(message_type, message_omf_json, action='create'):
-    # Sends the request out to the preconfigured endpoint..
+def send_message_to_omf_endpoint(endpoint, message_type, message_omf_json, action='create'):
+    '''Sends the request out to the preconfigured endpoint'''
 
-    global producerToken, omfEndPoint, omfVersion, sendingToOCS, username, password
     # Compress json omf payload, if specified
     compression = 'none'
-    if USE_COMPRESSION:
+    if endpoint["use-compression"]:
         msg_body = gzip.compress(bytes(json.dumps(message_omf_json), 'utf-8'))
         compression = 'gzip'
     else:
         msg_body = json.dumps(message_omf_json)
 
-    msg_headers = getHeaders(compression, message_type, action)
+    # Collect the message headers
+    msg_headers = get_headers(endpoint, compression, message_type, action)
+
+    # Send message to OMF endpoint
+    endpoints_type = endpoint["endpoint-type"]
     response = {}
-    # Assemble headers
-    if sendingToOCS:
+    # If the endpoint is OCS
+    if endpoints_type == EndpointTypes.OCS.value:
```

Comment:
Is there no concept of an enum switch statement in Python? I haven't seen it before in Python but usually this would be like
```
switch(endpoints_type):
  case EndpointTypes.OCS:
    ...do something...
```

Reply:
In most languages switch statements are reserved for integer values. In this case, I am using strings (despite also using an enum). If I were to use a switch statement I would likely have to add logic to get_config to parse the endpoint_type into an integer enumeration set. Is there a reason that a switch statement would be preferable here besides readability?

Reply:
It's only for readability. The languages I'm most familiar with (C#, TypeScript) allow strings or ints in switches. 
With that being said, it looks like Python doesn't have a built in switch at all, so you can ignore me :)
It looks like they might be added in Python 3.10: https://towardsdatascience.com/switch-case-statements-are-coming-to-python-d0caf7b2bfd3